subsys/net/ip/l2/wifi_shell.c

```diff
@@ -0,0 +1,244 @@
+#include <net/wifi_mgmt.h>
+#include <net/net_event.h>
+
+#define WIFI_SHELL_MODULE "wifi"
+
+#define WIFI_SHELL_MGMT_EVENTS (NET_EVENT_WIFI_SCAN_RESULT |		\
+				NET_EVENT_WIFI_SCAN_DONE |		\
+				NET_EVENT_WIFI_CONNECT_RESULT |		\
+				NET_EVENT_WIFI_DISCONNECT_RESULT)
+
+static union {
+	struct {
+		u8_t connecting		: 1;
+		u8_t disconnecting	: 1;
+		u8_t _unused		: 6;
+	};
+	u8_t all;
+} context;
+
+static u32_t scan_result;
+
+static struct net_mgmt_event_callback wifi_shell_mgmt_cb;
+
+static void handle_wifi_scan_result(struct net_mgmt_event_callback *cb)
+{
+	const struct wifi_scan_result *entry =
+		(const struct wifi_scan_result *)cb->info;
+
+	scan_result++;
+
+	if (scan_result == 1) {
+		printk("\n%-4s | %-32s %-5s | %-4s | %-4s | %-5s\n",
+		       "Num", "SSID", "(len)", "Chan", "RSSI", "Sec");
+	}
+
+	printk("%-4d | %-32s %-5u | %-4u | %-4d | %-5s\n",
+	       scan_result, entry->ssid, entry->ssid_length,
+	       entry->channel, entry->rssi,
+	       (entry->security == WIFI_SECURITY_TYPE_PSK ?
+		"WPA/WPA2" : "Open"));
+}
+
+static void handle_wifi_scan_done(struct net_mgmt_event_callback *cb)
+{
+	const struct wifi_status *status =
+		(const struct wifi_status *)cb->info;
+
+	if (status->status) {
+		printk("\nScan request failed (%d)\n", status->status);
+	} else {
+		printk("----------\n");
+		printk("Scan request done\n");
+	}
+
+	scan_result = 0;
+}
+
+static void handle_wifi_connect_result(struct net_mgmt_event_callback *cb)
+{
+	const struct wifi_status *status =
+		(const struct wifi_status *) cb->info;
+
+	if (status->status) {
+		printk("\nConnection request failed (%d)\n", status->status);
+	} else {
+		printk("\nConnected\n");
+	}
+
+	context.connecting = false;
+}
+
+static void handle_wifi_disconnect_result(struct net_mgmt_event_callback *cb)
+{
+	const struct wifi_status *status =
+		(const struct wifi_status *) cb->info;
+
+	if (context.disconnecting) {
+		printk("\nDisconnection request %s (%d)\n",
+		       status->status ? "failed" : "done",
+		       status->status);
+		context.disconnecting = false;
+	} else {
+		printk("\nDisconnected\n");
+	}
+}
+
+static void wifi_mgmt_event_handler(struct net_mgmt_event_callback *cb,
+				    u32_t mgmt_event, struct net_if *iface)
+{
+	switch (mgmt_event) {
+	case NET_EVENT_WIFI_SCAN_RESULT:
+		handle_wifi_scan_result(cb);
+		break;
+	case NET_EVENT_WIFI_SCAN_DONE:
+		handle_wifi_scan_done(cb);
+		break;
+	case NET_EVENT_WIFI_CONNECT_RESULT:
+		handle_wifi_connect_result(cb);
+		break;
+	case NET_EVENT_WIFI_DISCONNECT_RESULT:
+		handle_wifi_disconnect_result(cb);
+		break;
+	default:
+		break;
+	}
+}
+
+static int shell_cmd_connect(int argc, char *argv[])
+{
+	struct net_if *iface = net_if_get_default();
+	static struct wifi_connect_req_params cnx_params;
+	int idx = 0;
+
+	if (argc < 3) {
+		return -EINVAL;
+	}
+
+	cnx_params.ssid_length = atoi(argv[2]);
+	if (cnx_params.ssid_length <= 2) {
+		return -EINVAL;
+	}
+
+	cnx_params.ssid = &argv[1][1];
```

Comment:
This change keeps the first letter of the SSID from being stripped:
```
-       cnx_params.ssid = &argv[1][1];
+       cnx_params.ssid = argv[1];
```

Reply:
Hum yes that's done on purpose : ssid should be given starting with " and ending with ". see help  of the wifi shell module. It's because ssid can hold spaces.

Reply:
In an ideal implementation of the wifi shell, it would parse the SSID as a byte array. I did not bother doing that yet since neither simplelink nor asf treat ssid as a byte array as _they should_ because the wifi spec says so. But if we get a native wifi stack someday, it will be necessary.
As we don't store the results anywhere, it's not possible to use another information (such as a hash, or an index etc...). So length is needed.

Reply:
> I did not bother doing that yet since neither simplelink nor asf treat ssid as a byte array as they should because the wifi spec says so. 

OK.
Actually, SimpleLink connect API *does* treat the SSID as a byte array, and of a given length.
``_i16 sl_WlanConnect 	( 	const _i8 *  	pName,
		const _i16  	NameLen,
		const _u8 *  	pMacAddr,
		const SlWlanSecParams_t *  	pSecParams,
		const SlWlanSecParamsExt_t *  	pSecExtParams 
	) 	
``

Reply:
@tbursztyka, FYI, just got around to testing wifi_shell connecting to an AP with a space embedded in the SSID name.
Turns out, shell module line2arg() function doesn't support quotes.
`wifi> connect "HVL Network" 11 0 password`
yields tokens:
`{"HVL, Network", 11, 0, password}`
Filed https://github.com/zephyrproject-rtos/zephyr/issues/7178 , maybe the shell owner can add as an enhancement.